examples/framesjs-starter/app/examples/simplified-api/new-api.tsx

```diff
@@ -0,0 +1,286 @@
+import assert from "assert";
+import { ImageResponse } from "@vercel/og";
+import { ImageResponseOptions } from "next/server";
+import { getFrameMessage } from "frames.js/next/server";
+import { FrameActionDataParsedAndHubContext } from "frames.js";
+
+/** A subset of JS objects that are serializable */
+export type AnyJson = boolean | number | string | null | JsonArray | JsonMap;
+export interface JsonMap {
+  [key: string]: AnyJson;
+}
+interface JsonArray extends Array<AnyJson> {}
+
+interface IFrameImage {
+  renderToMetaTag(): Promise<string>;
+}
+
+export class Frame {
+  private image: IFrameImage | undefined;
+  private buttons: IFrameButton[] = [];
+  /**
+   * This is frame state coming from Button no the global app state comming from FramesApp
+   */
+  private state: AnyJson | undefined = undefined;
+
+  getState() {
+    return this.state;
+  }
+
+  setState(state: AnyJson | undefined) {
+    this.state = state;
+  }
+
+  registerImage(
+    definitionOrURL: React.ReactElement | string | URL
+  ): IFrameImage {
+    if (
+      typeof definitionOrURL === "object" &&
+      !(definitionOrURL instanceof URL)
+    ) {
+      return (this.image = new FrameImage(definitionOrURL));
+    }
+
+    return (this.image = new FrameImageURL(new URL(definitionOrURL)));
+  }
+
+  registerPostButton(path: `/${string}`, label: string): FramePostButton {
+    if (this.buttons.length >= 4) {
+      throw new Error("Maximum number of buttons is 4");
+    }
+
+    const button = new FramePostButton(
+      path,
+      (this.buttons.length + 1) as AllowedButtonIndexes,
+      label
+    );
+
+    this.buttons.push(button);
+
+    return button;
+  }
+
+  registerExternalLinkButton(href: URL, label: string): FrameExternalButton {
+    if (this.buttons.length >= 4) {
+      throw new Error("Maximum number of buttons is 4");
+    }
+
+    const button = new FrameExternalButton(
+      href,
+      (this.buttons.length + 1) as AllowedButtonIndexes,
+      label
+    );
+
+    this.buttons.push(button);
+
+    return button;
+  }
+
+  async toHTML(app: FramesApp): Promise<string> {
+    assert(this.image, "Frame image is not set");
+
+    return `
+      <!DOCTYPE html>
+      <html> 
+        <head>
+          <meta name="fc:frame" content="vNext" />
+          ${await this.image.renderToMetaTag()}
+          <meta name="fc:frame:state" content="${JSON.stringify(app.getState())}" />
+        </head>
+        <body></body>
+      </html>
+    `.trim();
+  }
+}
+
+type AllowedButtonIndexes = 1 | 2 | 3 | 4;
+
+interface IFrameButton {
+  renderToMetaTag(): string;
+}
+
+class FrameExternalButton implements IFrameButton {
+  private href: URL | undefined;
+  private index: AllowedButtonIndexes | undefined;
+  private label: string | undefined;
+
+  constructor(href: URL, index: AllowedButtonIndexes, label: string) {
+    this.href = href;
+    this.index = index;
+    this.label = label;
+  }
```

Comment:
Should try to stick more closely to the spec and avoid introducing new abstractions like href - buttons have an action and target

Reply:
Yep that makes sense.

Reply:
Ok when I'm thinking about it, it doesn't matter whether it is href or action, the point is, that it must be an URL: I don't like having disjoint unions to define stuff that much, it is more clear to have specific classes or functions to create specific buttons (e.g. `createMintButton`, `createPostButton` or something like this than `createButton({ action: 'post', ... })`).